astropy/io/fits/hdu/image.py

```diff
@@ -313,11 +369,30 @@ def update_header(self):
         if 'BLANK' in self._header:
             self._blank = self._header['BLANK']
 
+        # Add BSCALE/BZERO to header if data is unsigned int.
         self._update_uint_scale_keywords()
 
         self._modified = False
 
     def _update_header_scale_info(self, dtype=None):
+        """
+        Delete BSCALE/BZERO from header if necessary.
+        """
+
+        # Note that _dtype_for_bitpix determines the dtype based on the
+        # "original" values of bitpix, bscale, and bzero, stored in
+        # self._orig_bitpix, etc. It contains the logic for determining which
+        # special cases of BZERO/BSCALE, if any, are auto-detected as following
+        # the FITS unsigned int convention.
+
+        # Added original_was_unsigned with the intent of facilitating the
+        # special case of do_not_scale_image_data=True and uint=True
+        # eventually.
+        if self._dtype_for_bitpix() is not None:
+            original_was_unsigned = self._dtype_for_bitpix().kind == 'u'
+        else:
+            original_was_unsigned = False
```

Comment:
It looks like you were going to do something interesting here but haven't gotten around to it yet.  Maybe something to do with my comment [here](https://github.com/astropy/astropy/pull/5053/files#r66578803)?

Also regarding `dtype.kind == 'u'` something to be aware of is that this does not apply to 8-bit ints.  See #4267 which I believe is still not fixed, but is worth being aware of.